### ISSUE TITLE
How does NCCL know about internode topology when using PXN?

### ISSUE THREAD

**Issue by sezhiyanhari** [NVIDIA/nccl#918]
From this [blog post](https://developer.nvidia.com/blog/doubling-all2all-performance-with-nvidia-collective-communication-library-2-12/), one of the reasons for using PXN is to exploit rail topology. However, rail topology is an internode network topology, which to my understanding, NCCL is not aware of. To my understanding, NCCL is only aware of intranode topology, but please correct me if I'm wrong here.

If my assumption is correct, how exactly does NCCL know when to use PXN for the rail topology case?

**Comment by sjeaugey** [comment]
NCCL doesn't know and enables PXN always. That's because on top of helping on rail-optimized platforms, it also helps on any platforms, as it allows for message aggregation, dividing the number of IB/RoCE network messages by 8.

**Comment by sezhiyanhari** [comment]
Got it, that makes sense.

But in the case of rail-optimized platforms, how does PXN know which GPU to send data to for a given destination. Say for example GPU0 in node 0 is trying to send to GPU1 in node 1. If GPU1 in node 0 is connected to GPU1 in node 1 through a rail, how does PXN know to select GPU1 in node 0 as the proxy GPU?

**Comment by sezhiyanhari** [comment]
@sjeaugey would appreciate a response on this, thank you!

**Comment by KaimingOuyang** [comment]
You already said `GPU1 in node 0 is connected to GPU1 in node 1 through a rail`.

When it comes to rail-optimized, it means NICs close to the GPUs with same ID would connect to the same IB switch. NCCL utilizes this info to select GPU1 as proxy GPU instead of others.

**Comment by nikhilvgs** [comment]
Does NCCL assume the number of rails in the system is equal to the number of GPUs per node?

**Comment by nikhilvgs** [comment]
That makes sense. 

What happens if the number of rails is less than the number of NICs?
So, say, NIC is 1-1 with GPU in a node, but the network connectivity only ensures 4 rails, but connected consistently like
GPU0 => Rail0, GPU1 => Rail1, ... GPU4 => Rail0, ... GPU7 => Rail 3. 

Would NCCL still send traffic for GPU0 Node0 => GPU4 Node1  via GPU4 Node0?

**Comment by sjeaugey** [comment]
> Would NCCL still send traffic for GPU0 Node0 => GPU4 Node1 via GPU4 Node0?

Yes. As mentioned, NCCL cannot detect the network topology, so it just communicates preferably across same NICs. If those are on the same rail, it's great. Even if they're not, for alltoall operations you'd still want to have that extra step to allow for aggregation on GPU 4 of all 8 messages to remote GPU 4 so that we send a single NIC command.

**Comment by trstruth** [comment]
hey @sjeaugey is the info above still accurate? Wondering if nccl still does not attempt to detect the internode network topology

**Comment by sjeaugey** [comment]
Yes it's still accurate.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're [reviewing](https://github.com/NVIDIA/nccl/discussions/1761) older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.

**Comment by ButzinaDekardinta** [comment]
What about when using PXN with UCX? does UCX Mutli Rails - brake the PXN? Consider for example 2 hosts with 8 GPU each and 8 NIC each. So For example if GPU 6 on host A (Rank 6) decide to pass traffic to GPU 2 (Rank 10) on Host B, due to PXN the host will first pass the traffic from GPU 6 (Rank 6) via NVLink to GPU 2 (Rank 2) and then to NIC2 (in host A) and then via the Network Fabric to NIC 2 (in host B)  and then finally to GPU 2 (Rank 10). Since we have 1 Rail per nic (total 8 Rails) we used Rail 2 (assume there is no cross link in the cable and in the fabric, and assume we have enable the multi channel - so we will have multi tress to utilized all rails). To my understanding and this is my question, if you use UCX - they you brake the RAIL topology. As UCX might say in our example that it want to use NIC4 (in host A) due to dynamic load balancing or other factor NCCL doesn't know about (as things changed since bootstrap). So evven that PXN wanted to use NIC2 at the end UX used NIC4 - and we basically had unfair/unbalanced NVLink usage.
